### PR TITLE
Add support for the invisible ReCaptcha [MAILPOET-4145]

### PR DESCRIPTION
**Pull request by rodrigoprimo**
This PR adds support for the invisible ReCaptcha in the subscribe forms. Before, we only supported the checkbox ReCaptcha. The API is almost identical, but there are some differences. Because of those differences, it was necessary to change the code that handles the form submission on public.jsx a bit. 

For the invisible version, we need to call recaptcha.execute(), and then the ReCaptcha API calls the pre-defined callback set inside renderCaptcha() with a token if the captcha was solved successfully. The callback then calls submitSubscribeForm() with the token. For the checkbox version, we get the token synchronously inside the code that handles the form submission when recaptcha.getResponse() is called. So we can call submitSubscribeForm() directly.

I would suggest reviewing this PR commit by commit. The first commit makes the necessary changes to the admin to add a new setting to use the invisible ReCaptcha. The second commit just makes some organizational changes to public.jsx that are required for the next commit due to the no-use-before-define ESLint rule. Finally, the third commit changes the frontend code to display the invisible ReCaptcha if this is the selected setting.

**Testing instructions**

- Test that **submitting a form on the front-end works** for the three different types of captcha that we support (Built-in, Google reCAPTCHA v2 Checkbox, Google reCAPTCHA v2 Invisible) and also **when the captcha is disabled**.
- You will need to enter `Google ReCaptcha Keys` in `Settings` > `Advanced tab`... and find your keys in the `P2` doc
- Ensure that the **new setting in the admin works** according to the specs in the Jira ticket.
- It is a bit tricky to test that the invisible ReCaptcha is working as most of the time, it works without interfering with the form submission flow. You add a breakpoint to SubscriberSubscribeController::validateCaptcha() and confirm that this method is getting a ReCaptcha token and that the request that it makes to the ReCaptcha server has a successful response.
- Links to the ReCaptcha API documentation in the Jira ticket.

[MAILPOET-4145]

[MAILPOET-4145]: https://mailpoet.atlassian.net/browse/MAILPOET-4145?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: mailpoet/assets/js/src/public.jsx
Comment (triple0t) [inline mailpoet/assets/js/src/public.jsx:176]: From what I understand, this function `invisibleReCaptchaCallback` will be executed by the recaptcha API as a callback. 
Would these variables `form`, `formData` be available within the function scope?.

I can see this error in the console
Screenshot 2022-04-21 at 1 10 23 PM

When using the Invisible ReCaptcha option, submitting the form and solving the challenge.

File: mailpoet/assets/js/src/public.jsx
Comment (triple0t) [inline mailpoet/assets/js/src/public.jsx:400]: Could you reconfirm you are getting the right value for the `data-size` attribute here?

When using the Invisible ReCaptcha option, I was getting `size` as undefined when it should be 'invisible'

**Review by triple0t — COMMENTED** [review]
Awesome work @rodrigoprimo 

I tested and I can confirm the checkbox reCaptcha option works

But I was unable to get the Invisible reCaptcha option to work.

Here is the error from the UI:
Screenshot 2022-04-21 at 1 22 44 PM

Data gotten from the server:
Screenshot 2022-04-21 at 1 22 53 PM

POST request Form data:
Screenshot 2022-04-21 at 1 23 11 PM

POST request response:
Screenshot 2022-04-21 at 1 23 25 PM

Kindly check my comments.

Thank you

File: mailpoet/assets/js/src/settings/pages/advanced/captcha.tsx
Comment (rodrigoprimo) [reply mailpoet/assets/js/src/settings/pages/advanced/captcha.tsx:78]: I'm inclined to agree with you on this, but this is not something that I changed in this PR. Before the changes proposed here, we were already showing the "Sign up for an API key pair here" link when "Disable" is selected. Maybe we should address this in a separate ticket or discuss it with Jan M. before proceeding with this change here?

File: mailpoet/assets/js/src/public.jsx
Comment (rodrigoprimo) [reply mailpoet/assets/js/src/public.jsx:166]: I don't know why we are using compact as the default value for the checkbox ReCaptcha. This is a decision that was made before this PR, and probably before I joined MailPoet. Do you reckon this is something that we should change?

File: mailpoet/assets/js/src/public.jsx
Comment (rodrigoprimo) [reply mailpoet/assets/js/src/public.jsx:134]: I also noticed that we are calling window.grecaptcha.reset() twice and I'm not sure why. I opted to leave the code as is since this duplication was already present and, when creating this PR, I did not have time to investigate the reasons.

File: mailpoet/assets/js/src/public.jsx
Comment (rodrigoprimo) [reply mailpoet/assets/js/src/public.jsx:176]: > From what I understand, this function invisibleReCaptchaCallback will be executed by the recaptcha API as a callback.

That is correct.

>  Would these variables form, formData be available within the function scope?.

I believe you are referring only to `form` as `formData` is defined inside the callback function. I'm not a JS specialist so I could be missing something here, but I believe the answer is that `form` is available within the callback function because of the way that local scope works in JS. A variable defined inside a function is only accessible from that function and any nested functions.

> I can see this error in the console

I don't see this error in my environment. I made some changes related to another problem that you mentioned in the review comment. Could you please test again and check if you still get it?

File: mailpoet/assets/js/src/public.jsx
Comment (rodrigoprimo) [reply mailpoet/assets/js/src/public.jsx]: I was confused as well by those two variables when I first read the original code to create this PR, but I failed to act on it. Thanks for nudging me to take action and for the great suggestions for the new names for `formData.recaptcha` and `formData.data.recaptcha`. I just pushed a new commit implementing this change (208aa6fbefc4f91ccd6400b3692a2203ac6d79d2).

File: mailpoet/assets/js/src/public.jsx
Comment (rodrigoprimo) [reply mailpoet/assets/js/src/public.jsx]: See https://github.com/mailpoet/mailpoet/pull/4084#discussion_r859160836

**Comment by rodrigoprimo** [comment]
Thanks for the detailed review, @triple0t!

I believe I answered all your comments and this PR is ready for a second review.

> But I was unable to get the Invisible reCaptcha option to work.

I completely missed that different form templates had different HTML structures. I was testing with a form template different than the one you used to test and that explains why it was working for me, but not for you. I pushed a new commit (35b22a706c21fae58d678a1d60fcfc1d117ba761) to make the parts of the code that rely on the HTML structure more lenient and now the invisible ReCaptcha should work for the form template that you are using.

File: mailpoet/assets/js/src/public.jsx
Comment (rodrigoprimo) [reply mailpoet/assets/js/src/public.jsx]: See https://github.com/mailpoet/mailpoet/pull/4084#issuecomment-1110270323

File: mailpoet/assets/js/src/public.jsx
Comment (triple0t) [reply mailpoet/assets/js/src/public.jsx:166]: The major difference I noticed was the size of the captcha.

**Compact**
Screenshot 2022-04-27 at 4 15 08 PM

**Normal**
Screenshot 2022-04-27 at 4 20 37 PM

**No Captcha**
Screenshot 2022-04-27 at 4 21 08 PM


> This is a decision that was made before this PR, and probably before I joined MailPoet. Do you reckon this is something that we should change?

Without finding out the reason why we are using compact over normal, I believe we can leave it as it is.

File: mailpoet/assets/js/src/public.jsx
Comment (triple0t) [reply mailpoet/assets/js/src/public.jsx]: Thank you for performing the update 🙌🏿

**Review by triple0t — DISMISSED** [review]
Tested and working well.

Well done @rodrigoprimo and Thanks for updating the method names. 👏🏾

File: mailpoet/assets/js/src/settings/pages/advanced/captcha.tsx
Comment (pavel-mailpoet) [reply mailpoet/assets/js/src/settings/pages/advanced/captcha.tsx:78]: Yes please, only display the Sign up link when the reCAPTCHA is selected. The specification also asks to add:
"Please use the same reCAPTCHA type as in Google reCAPTCHA Settings." and I don't see that on the screenshot.

Comment (triple0t) [reply mailpoet/assets/js/src/settings/pages/advanced/captcha.tsx:78]: @pavel-mailpoet the "Please use the same reCAPTCHA type as in Google reCAPTCHA Settings." text shows up when any of the Google reCaptcha options is selected.
Screenshot 2022-04-28 at 8 26 45 AM

Comment (triple0t) [reply mailpoet/assets/js/src/settings/pages/advanced/captcha.tsx:78]: Hi @rodrigoprimo I'm Okay with the state of this PR. I'm going to re-assign it back to you. You can choose to move it ahead to QA or implement Pavels' suggestion.

Thank you 🙏

Comment (rodrigoprimo) [reply mailpoet/assets/js/src/settings/pages/advanced/captcha.tsx]: Thanks for your input, @pavel-mailpoet.

@triple0t, I added a new commit changing the logic that controls when the sign up link is visible. Now it should only appear when the user selects either ReCaptcha checkbox or ReCaptcha invisible. I'm assigning this PR back to you for a final review.

**Review by triple0t — DISMISSED** [review]
LGTM

File: mailpoet/assets/js/src/settings/pages/advanced/captcha.tsx
Comment (triple0t) [reply mailpoet/assets/js/src/settings/pages/advanced/captcha.tsx]: Approved. Thank you

**Comment by rohankacharegs** [comment]
GlobalStep Verified

**Review by veljkho — CHANGES_REQUESTED** [review]
Hi @rodrigoprimo 
I think we have issue with the new Invisible reCAPTCHa or at least not the good behavior...
It is the situation when the user receives some warning from reCAPTCHA v2 Invisible but unable to see it on the form itself... for example the warning "You need to wait X more minutes in order to subscribe again".. which is happening if you want to subscribe twice time in a row (even using different form) for a short amount of time (like 2 minutes).

Please check this GIF: https://d.pr/i/tyJteu
So, when clicking the Subscribe button nothing happens.. the error is missing here "you need to wait X more minutes..". On this page I have had the popup form where I was subscribed recently, and wanted to subscribe on this widget placed form also.. so probably the warning left on the popup form.

Is this fixable? Please let me know if you need more information regarding reproduction or  about the issue itself, would be glad to help.

**Comment by rodrigoprimo** [comment]
Good catch, @veljkho. I was not able to reproduce exactly the same setup that you had in the screencast that you shared, but I was able to reproduce a very similar bug.

If I'm not missing something, the original code was not resetting the ReCaptcha when there was an error in the form submission. Due to differences between the checkbox and invisible ReCaptchas, this meant that for the former there was an error but it didn't block the form submission. For the latter, the error did block the form submission which is the problem that you found.

I added a new commit to make sure we reset the ReCaptcha when there is an error with the form submission and in my testing, it fixed the problem that you found (or at least a very similar one, since I couldn't reproduce exactly the same setup).

Could you please test again?

(I'm rerunning CircleCI as a flaky acceptance test that is not related to this PR failed)

**Review by veljkho — CHANGES_REQUESTED** [review]
Hi Rodrigo!
I can still reproduce the issue. This is happening if you have more than 1 form on the same page... Please add 2 forms (can be 1 widget and 1 popup) on the same page and try subscribing to Popup first.. then reload the page and try subscribing to widget form..  you'll see that there's no warning message present, and that is because of the warning left on the popup instead showing on the widget form. Some conflicting is present here, the warnings left on previously submitted form. Could be this Google's issue?

**Comment by rodrigoprimo** [comment]
@veljkho, I missed from your previous comment that you were testing with two different form types. So the problem that I fixed is likely not the same as the one that you found. Apologies for that. I will ping you to get access to your online environment so that I can see exactly what you are seeing. Thanks.

**Comment by rodrigoprimo** [comment]
I chatted with @veljkho, and I understood the problem that he was referring to that I originally missed. What threw me off was that the screencast doesn't show that the problem happens when there are multiple forms on the page. Here is a screencast showing that when there are at least two forms on the page, and the invisible ReCaptcha is being used, the second form doesn't work as the code is wrongly submitting the first form:

https://user-images.githubusercontent.com/77215/167432283-c92a1ead-3482-468c-a862-5bc6602b5e2c.mp4

This was happening because of a bug in the code that I created that was always calling `grecaptcha.execute()` for the first form as I forgot to pass the ID of the ReCaptcha when calling this method. I missed this problem when testing because, while I did test with multiple forms on the same page, I tested only the first form, and this one works as the API uses its ID when no ID is passed to `grecaptcha.execute()`.

Here is a screencast of the second form working after the changes that I introduced in 1288929:

https://user-images.githubusercontent.com/77215/167432902-6d4c6deb-7aa9-4e72-b939-91af8bd5cd89.mp4

@veljkho, could you please check again? We might want @triple0t to check this PR one more time just to be on the safe side, as I introduced a few code changes after his review.

**Comment by rohankacharegs** [comment]
GlobalStep Verified

**Review by veljkho — APPROVED** [review]

**Comment by veljkho** [comment]
It is working now @rodrigoprimo thanks! :)
Please fix conflicts so I can merge it.

**Comment by rodrigoprimo** [comment]
@veljkho, I fixed the conflicts, but I had to use rebase instead of merge, and that means that I need another approval review before we can merge this PR. Could you please check it again?

**Review by veljkho — APPROVED** [review]

**Pull request merged**
